lib/semantics/resolve-names.cc

```diff
@@ -1037,11 +1037,271 @@ class ConstructVisitor : public DeclarationVisitor {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Firstprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpFirstPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Lastprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpLastPrivate);
+    return false;
+  }
+
+  void add_dataSharingAttributeObject(const Symbol *object) {
+    dataSharingAttributeObjects_.insert(object);
+  }
+  void clear_dataSharingAttributeObject() {
+    dataSharingAttributeObjects_.clear();
+  }
+  bool find_dataSharingAttributeObject(const Symbol *);
+
+protected:
+  // TODO: resolve variables referenced in the OpenMP region
+  void ResolveOmpObjectList(const parser::OmpObjectList &, Symbol::Flag);
+  void ResolveOmpObject(const parser::OmpObject &, Symbol::Flag);
+  Symbol *ResolveOmp(const parser::Name &, Symbol::Flag);
+  Symbol *ResolveOmp(Symbol &, Symbol::Flag);
+  Symbol *ResolveOmpCommonBlockName(const parser::Name *);
+  Symbol *DeclarePrivateAccessEntity(const parser::Name &, Symbol::Flag);
+  Symbol *DeclarePrivateAccessEntity(Symbol &, Symbol::Flag);
+  Symbol *DeclareOrMarkOtherAccessEntity(const parser::Name &, Symbol::Flag);
+  Symbol *DeclareOrMarkOtherAccessEntity(Symbol &, Symbol::Flag);
+  void CheckMultipleAppearances(
+      const parser::Name &, const Symbol *, Symbol::Flag);
+
+private:
+  std::set<const Symbol *> dataSharingAttributeObjects_;  // on one directive
+};
+
+bool OmpVisitor::find_dataSharingAttributeObject(const Symbol *object) {
+  auto it{dataSharingAttributeObjects_.find(object)};
+  if (it != dataSharingAttributeObjects_.end()) {
+    return true;
+  }
+  return false;
+}
+
+Symbol *OmpVisitor::ResolveOmpCommonBlockName(const parser::Name *name) {
+  if (auto *prev{name ? currScope().parent().FindCommonBlock(name->source)
+                      : nullptr}) {
+    name->symbol = prev;
+    return prev;
+  } else {
+    return nullptr;
+  }
+}
+
+void OmpVisitor::ResolveOmpObjectList(
+    const parser::OmpObjectList &ompObjectList, Symbol::Flag ompFlag) {
+  for (const auto &ompObject : ompObjectList.v) {
+    ResolveOmpObject(ompObject, ompFlag);
+  }
+}
+
+void OmpVisitor::ResolveOmpObject(
+    const parser::OmpObject &ompObject, Symbol::Flag ompFlag) {
+  if (const auto *designator{std::get_if<parser::Designator>(&ompObject.u)}) {
+    if (const auto *name{GetDesignatorNameIf(*designator)}) {
+      auto *symbol{ResolveOmp(*name, ompFlag)};
+      if (dataSharingAttributeFlags.test(ompFlag)) {
+        CheckMultipleAppearances(*name, symbol, ompFlag);
+      }
+    } else if (const auto *designatorName{ResolveDesignator(*designator)};
+               designatorName->symbol) {
+      // Array sections to be changed to substrings as needed
+      if (AnalyzeExpr(context(), *designator)) {
+        if (const auto *substring{
+                std::get_if<parser::Substring>(&designator->u)}) {
```

Comment:
`if (std::holds_alternative<parser::Substring>(designator->u))`

I'm surprised you don't get a warning for the unused variable `substring`.

Reply:
Thanks for the detailed suggestion. It surprised me too since you mentioned...